dvc/objects.py

```diff
@@ -0,0 +1,142 @@
+import logging
+
+from .exceptions import DvcException
+
+logger = logging.getLogger(__name__)
+
+
+class ObjectError(DvcException):
+    pass
+
+
+class ObjectFormatError(ObjectError):
+    pass
+
+
+class HashFile:
+    def __init__(self, path_info, tree, hash_info):
+        self.path_info = path_info
+        self.tree = tree
+        self.hash_info = hash_info
+
+    def __len__(self):
+        return 1
+
+    def __str__(self):
+        return f"object {self.hash_info}"
+
+    def __bool__(self):
+        return bool(self.hash_info)
+
+    def check(self):
+        actual = self.tree.get_hash(self.path_info)
+
+        logger.trace(
+            "cache '%s' expected '%s' actual '%s'",
+            self.path_info,
+            self.hash_info,
+            actual,
+        )
+
+        assert actual.name == self.hash_info.name
+        if actual.value.split(".")[0] != self.hash_info.value.split(".")[0]:
+            raise ObjectFormatError(f"{self} is corrupted")
+
+    @classmethod
+    def load(cls, odb, hash_info):
+        return odb.get(hash_info)
+
+    def commit(self, odb, **kwargs):
+        odb.add(self, **kwargs)
+
+
+class File(HashFile):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.src = None
+
+    @classmethod
+    def stage(cls, odb, path_info, tree, **kwargs):
+        hash_info = tree.get_hash(path_info, **kwargs)
+        raw = odb.get(hash_info)
+        obj = cls(raw.path_info, raw.tree, hash_info)
+        obj.src = HashFile(path_info, tree, hash_info)
+        return obj
+
+    def commit(self, odb, **kwargs):
+        self.src.commit(odb, **kwargs)
+
+
+class Tree(HashFile):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.src_hash_info = None
+        self.src_path_info = None
+        self.src_tree = None
+
+    def __len__(self):
+        return self.hash_info.dir_info.nfiles
+
+    def items(self, **kwargs):
+        for (
+            key,
+            hash_info,
+        ) in self.hash_info.dir_info.trie.iteritems(  # noqa: B301
+            **kwargs
+        ):
+            yield (key, hash_info)
+
+    @classmethod
+    def stage(cls, odb, path_info, tree, **kwargs):
```

Comment:
Similar to File, we are using a lazy approach for files, but actually save the '.dir' cache file, same as it was before.